surface/include/pcl/surface/impl/mls.hpp

```diff
@@ -673,45 +534,363 @@ pcl::MovingLeastSquares<PointInT, PointOutT>::performUpsampling (PointCloudOut &
+  pcl::computeCovarianceMatrix (cloud, nn_indices, xyz_centroid, covariance_matrix);
+  EIGEN_ALIGN16 Eigen::Vector3d::Scalar eigen_value;
+  EIGEN_ALIGN16 Eigen::Vector3d eigen_vector;
+  Eigen::Vector4d model_coefficients;
+  pcl::eigen33 (covariance_matrix, eigen_value, eigen_vector);
+  model_coefficients.head<3> ().matrix () = eigen_vector;
+  model_coefficients[3] = 0;
+  model_coefficients[3] = -1 * model_coefficients.dot (xyz_centroid);
+
+  // Projected query point
+  mls_result.valid = true;
+  mls_result.query_point = cloud.points[index].getVector3fMap ().template cast<double> ();
+  double distance = mls_result.query_point.dot (model_coefficients.head<3> ()) + model_coefficients[3];
+  mls_result.mean = mls_result.query_point - distance * model_coefficients.head<3> ();
+
+  mls_result.curvature = static_cast<float> (covariance_matrix.trace ());
+  // Compute the curvature surface change
+  if (mls_result.curvature != 0)
+    mls_result.curvature = fabsf (float (eigen_value / double (mls_result.curvature)));
+
+  // Get a copy of the plane normal easy access
+  mls_result.plane_normal = model_coefficients.head<3> ();
+
+  // Local coordinate system (Darboux frame)
+  mls_result.v_axis = mls_result.plane_normal.unitOrthogonal ();
+  mls_result.u_axis = mls_result.plane_normal.cross (mls_result.v_axis);
+
+  // Perform polynomial fit to update point and normal
+  ////////////////////////////////////////////////////
+  mls_result.num_neighbors = static_cast<int> (nn_indices.size ());
+  mls_result.polynomial_fit = polynomial_fit;
+  if (polynomial_fit)
+  {
+    mls_result.order = polynomial_order;
+    int nr_coeff = (polynomial_order + 1) * (polynomial_order + 2) / 2;
+
+    if (mls_result.num_neighbors >= nr_coeff)
+    {
+      // Note: The max_sq_radius parameter is only used if weight_func was not defined
+      double max_sq_radius = 1;
+      if (weight_func == 0)
+      {
+        max_sq_radius = search_radius * search_radius;
+        weight_func = boost::bind (&pcl::computeMLSWeight, _1 , max_sq_radius);
+      }
+
+      // Allocate matrices and vectors to hold the data used for the polynomial fit
+      Eigen::VectorXd weight_vec (mls_result.num_neighbors);
+      Eigen::MatrixXd P (nr_coeff, mls_result.num_neighbors);
+      Eigen::VectorXd f_vec (mls_result.num_neighbors);
+      Eigen::MatrixXd P_weight; // size will be (nr_coeff_, nn_indices.size ());
+      Eigen::MatrixXd P_weight_Pt (nr_coeff, nr_coeff);
+
+      // Update neighborhood, since point was projected, and computing relative
+      // positions. Note updating only distances for the weights for speed
+      std::vector<Eigen::Vector3d, Eigen::aligned_allocator<Eigen::Vector3d> > de_meaned (mls_result.num_neighbors);
+      for (size_t ni = 0; ni < mls_result.num_neighbors; ++ni)
+      {
+        de_meaned[ni][0] = cloud.points[nn_indices[ni]].x - mls_result.mean[0];
+        de_meaned[ni][1] = cloud.points[nn_indices[ni]].y - mls_result.mean[1];
```

Comment:
I'm not sure vectorization is properly exploited here. You should use the vector maps for this operation.

Reply:
Sorry, I don't follow. Can you provide more detail?